qbraid/interface/qbraid_qasm/tools.py

```diff
@@ -54,3 +56,99 @@ def _convert_to_contiguous_qasm(qasmstr: str, rev_qubits=False) -> QASMType:
 def _unitary_from_qasm(qasmstr: QASMType) -> np.ndarray:
     """Return the unitary of the QASM"""
     return from_qasm(qasmstr).unitary()
+
+
+def _build_qasm_3_reg(line: str, qreg_type: bool) -> QASMType:
+    """Helper function to build openqasm 3 register statements
+
+    Args:
+        line (str): openqasm 2  regdecl statement
+        qreg_type (bool): whether a qreg or creg type statement
+
+    Returns:
+        str : openqasm 3 qubits / bits declaration
+    """
+    reg_keyword_len = 4
+    line = line[reg_keyword_len:]
+    elements = line.split("[")
+    reg_name = elements[0].strip()
+    reg_size = elements[1].split("]")[0].strip()
+    result = "qubit" if qreg_type else "bit"
+    result += f"[{reg_size}] {reg_name};\n"
+    return result
+
+
+def _build_qasm_3_measure(line: str) -> QASMType:
+    """Helper function to build openqasm 3 measure string
+
+    Args:
+        line (str): openqasm 2 measure statement
+
+    Returns:
+        str:  openqasm 3 measure statement
+    """
+    measure_keyword_len = 7
+    line = line[measure_keyword_len:]
+    elements = line.split("->")
+    qubits_name = elements[0].replace(" ", "")
+    bits_name = elements[1].split(";")[0].replace(" ", "")
+
+    return f"{bits_name} = measure {qubits_name};\n"
+
+
+def _change_to_qasm_3(line: str) -> QASMType:
+    """Function to change an openqasm 2 line to openqasm 3
+
+    Args:
+        line (str): an openqasm 2 line
+
+    Returns:
+        str: corresponding openqasm 3 line
+    """
+    line = line.lstrip()
+    if line.startswith("OPENQASM"):
+        return ""
+    if "qelib1.inc" in line:
+        return ""
+    if line.startswith("qreg"):
+        return _build_qasm_3_reg(line, qreg_type=True)
+    if line.startswith("creg"):
+        return _build_qasm_3_reg(line, qreg_type=False)
+    if line.startswith("measure"):
+        return _build_qasm_3_measure(line)
+    if line.startswith("opaque"):
+        # as opaque is ignored by openqasm 3 add it as a comment
+        return "// " + line + "\n"
+    return line + "\n"
+
+
+def convert_to_qasm_3(qasm_2_str: str):
+    """Convert a QASM 2.0 string to QASM 3.0 string
+
+    Args:
+        qasm_2_str (str): QASM 2.0 string
+    """
+    try:
+        # use inbuilt method to check validity
+        _ = QuantumCircuit.from_qasm_str(qasm_2_str)
```

Comment:
If we've implemented our conversion function correctly, we shouldn't need qiskit's help identifying invalid qasm strings. Instead, our function should have it's own error handling to identify and raise exceptions when things aren't right! As a bonus for catching our own errors, it's one less import at the top, and we cut down the runtime by 1/2 (one pass over the circuit instead of two, which doesn't make a difference for small circuits, but is good practice especially if used in other functions that perform subsequent conversions)

Reply:
Yes, that does seem a good approach @ryanhill1 but then the current line-by-line parsing method would not work. If we were to catch _arbitrary errors_ in qasm2 strings, a parser would be required for the conversion which can report those.

I saw that there's already one built as the `QasmParser` class in `qbraid/transpiler/cirq_qasm/qasm_parser.py` which could be tweaked to build a qasm3 string on-the-fly but I wanted to first confirm if that would indeed be required.

It is also highly probable that I'm missing something and the parser is not required after all!

Reply:
That's a fair point. Since we're more swapping operators than parsing, I guess the error handling would be less straight forward. So for now this approach is ok. But going forward we'll likely integrate it in with our internal `QasmParser`, as we don't want to be dependent on qiskit to determine what can or can't be converted.

As one simple example, let's say for whatever reason qiskit's  `from_qasm_str` method is updated and has a bug. Then, it's possible that even for a valid qasm string, the qiskit check could fail and raise an exception, when the conversion would have worked on our end. But like I mentioned, we can address this later on when we integrate it. Thanks for your thoughts!